pulser-simulation/pulser_simulation/simresults.py

```diff
@@ -16,26 +16,35 @@
 from __future__ import annotations
 
 import collections.abc
+import typing
 from abc import ABC, abstractmethod
 from collections import Counter
 from functools import lru_cache
-from typing import Mapping, Optional, Tuple, Union, cast
+from typing import Mapping, Optional, Tuple, TypeVar, Union, cast, overload
 
 import matplotlib.pyplot as plt
 import numpy as np
 import qutip
 from numpy.typing import ArrayLike
 from qutip.piqs import isdiagonal
 
+from pulser.result import Result, SampledResult
+from pulser_simulation.qutip_result import QutipResult
 
-class SimulationResults(ABC):
+ResultType = TypeVar("ResultType", bound=Result)
+
+
+class SimulationResults(ABC, typing.Sequence[ResultType]):
```

Comment:
So now the `SimulationsResults` will be the return type for all the backends? And each backend will have, instead, a different `Result` class implementation. Is this the idea? If so, I think it works fine. I would have hoped to remove this additional `SimulationResults` and just have a single `Result` class (implemented differently for each backend) but I guess it was not really possible.

Reply:
Yes, perfect. This won't be an issue if the conceptual difference is well documented. However, I surmise we might also have, at some point, results at each evaluation time for our emulators. How would you treat that? In the end, I guess you could just do a `Sequence[Sequence[Result]]` and keep the generic result more flexible. It won't be a big change.